### DISCUSSION TITLE
Should a generic type be inferred to `Any` when its `TypeVar` is bound?

### DISCUSSION THREAD

**Discussion by augustwester** [microsoft/pyright#7243] [Q&A] [answered]
Suppose I have the following class with a bound generic type:

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

@dataclass
Metadata:
    uuid: str

T = TypeVar("T", bound=Metadata)

class Data(Generic[T]):
    metadata: T
```

I then have a function which returns instances of `Data`, possibly using subtypes of `Metadata`:

```python
def get_data() -> Data:
    # instantiate Data object
    return data

my_data = get_data()
metadata = my_data.metadata # <-- type is Any here?
```

Is this intended behavior? I would think that since `T` is bound to `Metadata`, the type checker should be able to infer that `metadata` is of type `Metadata`, so I will see `uuid` in my code suggestions when writing `my_data.metadata.`.

**Comment by erictraut** [comment, marked as answer]
Yes, this is the correct behavior as specified in the typing spec. You've defined a generic class called `Data` that takes a single type argument, but you've omitted this type argument when annotating the return type for `get_data`. The typing spec indicates that type checkers should assume that the type argument is implicitly `Any` in this case. The fact that you've added an explicit upper bound for the type parameter doesn't matter. (Note: if you don't provide an explicit upper bound, a type parameter still has an upper bound of `object`.)

There is a draft [PEP 696](https://peps.python.org/pep-0696/) that allows you to specify a "default value" for a TypeVar. Pyright has experimental support for this facility. If you set `enableExperimentalFeatures` to true in your pyright configuration, you can do the following:

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

@dataclass
class Metadata:
    uuid: str

T = TypeVar("T", bound=Metadata, default=Metadata)

class Data(Generic[T]):
    metadata: T

def get_data() -> Data: ...

my_data = get_data()
metadata = my_data.metadata  # <-- type is Metadata here
```

Or, using the new Python 3.12 syntax (which PEP 696 proposes to extend in Python 3.13 to support defaults), it would look like this:

```python
class Data[T: Metadata = Metadata]:
    metadata: T
```

**Reply by erictraut** [reply]
You might also be interested in enabling pyright's `reportMissingTypeArgument` option. It will tell you if you have omitted a type argument when the corresponding type parameter has no default — and therefore defaults to `Any`.
